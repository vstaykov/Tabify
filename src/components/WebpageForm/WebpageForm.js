import React from "react";
import PropTypes from "prop-types";

const WebpageForm = ({
  enabled,
  handleSubmitForm,
  pageUrl,
  handlePageUrlChange,
  pinned,
  handlePinnedChange,
  invalidDataMessage
}) => (
  <form onSubmit={handleSubmitForm}>
    <div>
      <label htmlFor="pageUrl">
        Page URL
        <input
          id="pageUrl"
          type="text"
          value={pageUrl}
          onChange={handlePageUrlChange}
        />
      </label>
      <label htmlFor="pinned" className="switch">
        Pinned
        <input
          id="pinned"
          type="checkbox"
          checked={pinned}
          onChange={handlePinnedChange}
        />
        <span className="slider round" />
      </label>

      <input type="submit" value="Add" disabled={!enabled} />
    </div>
    {invalidDataMessage && (
      <div className="invalid-form-data">{invalidDataMessage}</div>
    )}
  </form>
);

WebpageForm.propTypes = {
  enabled: PropTypes.bool.isRequired,
  handleSubmitForm: PropTypes.func.isRequired,
  pageUrl: PropTypes.string.isRequired,
  handlePageUrlChange: PropTypes.func.isRequired,
  pinned: PropTypes.bool.isRequired,
  handlePinnedChange: PropTypes.func.isRequired,
  invalidDataMessage: PropTypes.string.isRequired
};

export default WebpageForm;
