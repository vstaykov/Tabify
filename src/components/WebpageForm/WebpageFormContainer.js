import React from "react";
import PropTypes from "prop-types";
import WebPageForm from "./WebpageForm";

const URLREGEX = /^(ftp|http|https):\/\/[^ "]+$/;
const INVALIDURLMESSAGE =
  "Please provide a valid URL in the format ftp|http|https://<url with no spaces or quotes>";

class WebpageFormContainer extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      pageUrl: "",
      pinned: false,
      urlDataIsValid: true
    };
  }

  handlePageUrlChange = event => {
    this.setState({ pageUrl: event.target.value });
    event.preventDefault();
  };

  handlePinnedChange = event => {
    this.setState({ pinned: event.target.checked });
  };

  handleSubmitForm = event => {
    event.preventDefault();
    event.stopPropagation();

    if (this.validateUrlData()) {
      this.props.submit(this.state.pageUrl, this.state.pinned);
      this.setState({
        pageUrl: "",
        pinned: false,
        urlDataIsValid: true
      });
    } else {
      this.setState({
        urlDataIsValid: false
      });
    }
  };

  validateUrlData = () => {
    const urlIsValid = URLREGEX.test(this.state.pageUrl);
    return urlIsValid;
  };

  render() {
    const { pageUrl, pinned, urlDataIsValid } = this.state;
    const { enabled } = this.props;
    const invalidDataMessage = urlDataIsValid ? "" : INVALIDURLMESSAGE;

    return React.createElement(WebPageForm, {
      enabled,
      handleSubmitForm: this.handleSubmitForm,
      pageUrl,
      handlePageUrlChange: this.handlePageUrlChange,
      pinned,
      handlePinnedChange: this.handlePinnedChange,
      invalidDataMessage
    });
  }
}

WebpageFormContainer.propTypes = {
  submit: PropTypes.func.isRequired,
  enabled: PropTypes.bool.isRequired
};

export default WebpageFormContainer;
