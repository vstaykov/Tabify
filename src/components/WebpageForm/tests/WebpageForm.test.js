import React from "react";
import { shallow } from "enzyme";
import WebpageForm from "./../WebpageForm";

describe("<WebpageForm />", () => {
  let props;

  beforeEach(() => {
    props = {
      enabled: true,
      handleSubmitForm: () => {},
      pageUrl: "foo",
      handlePageUrlChange: () => {},
      pinned: true,
      handlePinnedChange: () => {},
      invalidDataMessage: ""
    };
  });

  describe("when invalid data message not provided", () => {
    it("should not render invalid data message section", () => {
      props.invalidDataMessage = "";

      const component = shallow(<WebpageForm {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe("when invalid data message provided", () => {
    it("should render invalid data message", () => {
      props.invalidDataMessage = "data is invalid";

      const component = shallow(<WebpageForm {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe("when enabled is true", () => {
    it("should render submit input enabled", () => {
      props.enabled = true;

      const component = shallow(<WebpageForm {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe("when enabled is false", () => {
    it("should render submit input disabled", () => {
      props.enabled = false;

      const component = shallow(<WebpageForm {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe("when pinned is true", () => {
    it("should render pinned checkbox checked", () => {
      props.pinned = true;

      const component = shallow(<WebpageForm {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe("when pinned is false", () => {
    it("should render pinned checkbox unchecked", () => {
      props.pinned = false;

      const component = shallow(<WebpageForm {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe("when page URL has initial value", () => {
    it("should render the value in page URL input", () => {
      props.pageUrl = "URL";

      const component = shallow(<WebpageForm {...props} />);

      expect(component).toMatchSnapshot();
    });
  });

  describe("when form submitted", () => {
    it("should call handleSubmitForm()", () => {
      props.handleSubmitForm = jest.fn();

      const component = shallow(<WebpageForm {...props} />);
      const form = component.find("form");

      form.simulate("submit");

      expect(props.handleSubmitForm).toHaveBeenCalledTimes(1);
    });
  });

  describe("when pinned changes", () => {
    it("should call handlePinnedChange()", () => {
      props.pinned = true;
      props.handlePinnedChange = jest.fn();

      const component = shallow(<WebpageForm {...props} />);
      const pinnedCheckbox = component.find("#pinned");

      pinnedCheckbox.simulate("change", { target: { checked: false } });
      pinnedCheckbox.simulate("change", { target: { checked: true } });

      expect(props.handlePinnedChange).toBeCalledTimes(2);
    });
  });

  describe("when page URL changes", () => {
    it("should call handlePageUrlChange()", () => {
      props.pageUrl = "";
      props.handlePageUrlChange = jest.fn();

      const component = shallow(<WebpageForm {...props} />);
      const pageUrlInput = component.find("#pageUrl");

      pageUrlInput.simulate("change", { target: { value: "foo" } });
      pageUrlInput.simulate("change", { target: { value: "bar" } });

      expect(props.handlePageUrlChange).toBeCalledTimes(2);
    });
  });
});
