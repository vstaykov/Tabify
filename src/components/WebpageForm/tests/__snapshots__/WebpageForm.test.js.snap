// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WebpageForm /> when enabled is false should render submit input disabled 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      htmlFor="pageUrl"
    >
      Page URL
      <input
        id="pageUrl"
        onChange={[Function]}
        type="text"
        value="foo"
      />
    </label>
    <label
      className="switch"
      htmlFor="pinned"
    >
      Pinned
      <input
        checked={true}
        id="pinned"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider round"
      />
    </label>
    <input
      disabled={true}
      type="submit"
      value="Add"
    />
  </div>
</form>
`;

exports[`<WebpageForm /> when enabled is true should render submit input enabled 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      htmlFor="pageUrl"
    >
      Page URL
      <input
        id="pageUrl"
        onChange={[Function]}
        type="text"
        value="foo"
      />
    </label>
    <label
      className="switch"
      htmlFor="pinned"
    >
      Pinned
      <input
        checked={true}
        id="pinned"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider round"
      />
    </label>
    <input
      disabled={false}
      type="submit"
      value="Add"
    />
  </div>
</form>
`;

exports[`<WebpageForm /> when invalid data message not provided should not render invalid data message section 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      htmlFor="pageUrl"
    >
      Page URL
      <input
        id="pageUrl"
        onChange={[Function]}
        type="text"
        value="foo"
      />
    </label>
    <label
      className="switch"
      htmlFor="pinned"
    >
      Pinned
      <input
        checked={true}
        id="pinned"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider round"
      />
    </label>
    <input
      disabled={false}
      type="submit"
      value="Add"
    />
  </div>
</form>
`;

exports[`<WebpageForm /> when invalid data message provided should render invalid data message 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      htmlFor="pageUrl"
    >
      Page URL
      <input
        id="pageUrl"
        onChange={[Function]}
        type="text"
        value="foo"
      />
    </label>
    <label
      className="switch"
      htmlFor="pinned"
    >
      Pinned
      <input
        checked={true}
        id="pinned"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider round"
      />
    </label>
    <input
      disabled={false}
      type="submit"
      value="Add"
    />
  </div>
  <div
    className="invalid-form-data"
  >
    data is invalid
  </div>
</form>
`;

exports[`<WebpageForm /> when page URL has initial value should render the value in page URL input 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      htmlFor="pageUrl"
    >
      Page URL
      <input
        id="pageUrl"
        onChange={[Function]}
        type="text"
        value="URL"
      />
    </label>
    <label
      className="switch"
      htmlFor="pinned"
    >
      Pinned
      <input
        checked={true}
        id="pinned"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider round"
      />
    </label>
    <input
      disabled={false}
      type="submit"
      value="Add"
    />
  </div>
</form>
`;

exports[`<WebpageForm /> when pinned is false should render pinned checkbox unchecked 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      htmlFor="pageUrl"
    >
      Page URL
      <input
        id="pageUrl"
        onChange={[Function]}
        type="text"
        value="foo"
      />
    </label>
    <label
      className="switch"
      htmlFor="pinned"
    >
      Pinned
      <input
        checked={false}
        id="pinned"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider round"
      />
    </label>
    <input
      disabled={false}
      type="submit"
      value="Add"
    />
  </div>
</form>
`;

exports[`<WebpageForm /> when pinned is true should render pinned checkbox checked 1`] = `
<form
  onSubmit={[Function]}
>
  <div>
    <label
      htmlFor="pageUrl"
    >
      Page URL
      <input
        id="pageUrl"
        onChange={[Function]}
        type="text"
        value="foo"
      />
    </label>
    <label
      className="switch"
      htmlFor="pinned"
    >
      Pinned
      <input
        checked={true}
        id="pinned"
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="slider round"
      />
    </label>
    <input
      disabled={false}
      type="submit"
      value="Add"
    />
  </div>
</form>
`;
